src/ide/code_lens/tests.rs

```diff
@@ -197,69 +191,119 @@ impl TestRunner {
-                        },
-                        false,
-                        file_state,
-                    );
-                };
-            }
-            ModuleItem::Module(module) => {
-                if let MaybeModuleBody::Some(body) = module.body(db) {
-                    let tests_count = file_state.len();
-
-                    collect_tests(
-                        db,
-                        body.items(db).elements(db),
-                        file,
-                        file_url.clone(),
-                        file_state,
-                    );
-
-                    // Append mod only if it contains tests.
-                    if tests_count != file_state.len() {
-                        if let Some(position) = module
-                            .as_syntax_node()
-                            .span_start_without_trivia(db)
-                            .position_in_file(db, file)
-                        {
-                            let position = position.to_lsp();
-
-                            make_code_lens(
-                                TextDocumentPositionParams {
-                                    position,
-                                    text_document: TextDocumentIdentifier { uri: file_url.clone() },
-                                },
-                                true,
-                                file_state,
-                            );
-                        };
-                    }
+    collect_functions(db, module, make_function_code_lens(db, origin_file, &file_url, file_state));
+
+    let Ok(modules) = db.module_submodules_ids(module) else { return };
+
+    for submodule in modules.iter().copied() {
+        let has_tests = if db.is_submodule_inline(submodule) {
+            let tests_count = file_state.len();
+
+            collect_tests(
+                db,
+                ModuleId::Submodule(submodule),
+                origin_file,
+                file_url.clone(),
+                file_state,
+            );
+
+            // Append mod only if it contains tests.
+            tests_count != file_state.len()
+        } else {
+            has_any_test(db, ModuleId::Submodule(submodule))
+        };
+
+        if !has_tests {
+            continue;
+        }
+
+        let ptr = submodule.stable_ptr(db).untyped();
+
+        let Some(position) = get_position(db, origin_file, ptr) else {
+            continue;
+        };
+
+        if ptr.file_id(db) != origin_file {
+            continue;
+        }
+
+        make_code_lens(
+            TextDocumentPositionParams {
+                position,
+                text_document: TextDocumentIdentifier { uri: file_url.clone() },
+            },
+            true,
+            file_state,
+        );
+    }
+}
+
+fn has_any_test(db: &AnalysisDatabase, module: ModuleId) -> bool {
+    if collect_functions(db, module, |_| ControlFlow::Break(())).is_break() {
```

Comment:
wat does it do

it feels like this function does 2 different things at once

I know it's less code, but it's less readable as well